File: Design2_HashSet.cpp
// Time Complexity : O(1)
// Space Complexity : O(1000*1000)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach

class MyHashSet {
    vector<vector<bool>> storage;
    int buckets;
    int bucketItems;
    int bucket(int key){
        return key%buckets;
    }
    int bucketItem(int key){
        return key/bucketItems;
    }
public:
    /** Initialize your data structure here. */
    MyHashSet() {
        buckets=1000;
        bucketItems=1000;
        storage.resize(buckets);
        //vector<vector<int>> storage(buckets); 
        //storage=new bool[buckets];
    }
    void add(int key) {
        int a=bucket(key);
        int b=bucketItem(key);
        if(storage[a].empty()){
            storage[a]=vector<bool> (bucketItems);
        }
        storage[a][b]= true;
    }
    void remove(int key) {
        int a=bucket(key);
        int b=bucketItem(key);
        if(storage[a].empty()) return;
        storage[a][b]=false;
    }
    /** Returns true if this set contains the specified element */
    bool contains(int key) {
        int a=bucket(key);
        int b=bucketItem(key);
        if(storage[a].empty()) return false;
        return storage[a][b];
    }
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet* obj = new MyHashSet();
 * obj->add(key);
 * obj->remove(key);
 * bool param_3 = obj->contains(key);
 */

File: Design-2 _HashSet.java
// Time Complexity : O(1)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


import java.util.*;
//I have declared size for hash box which is max_val and array size
//one parent list
// approach is to generate parent arraylist and assigning index using key % arraysize which 1000000 
class MyHashSet {
     private final int Max_Value = 1000000;
     private final int array_size = 100;
    private  List<List<Integer>> parentList;
    /** Initialize your data structure here. */
    public MyHashSet() {
        parentList = new ArrayList<>(array_size);
        for(int i =0; i<array_size; i++){
            parentList.add(null);
        }
    }
    //childlist is empty the linkedlist is added with that key and in prentlist the ky and index is updated and if the key is not present in childlist, we need to add it
    public void add(int key) {
        int index = key % array_size;
        List<Integer> childList = parentList.get(index);
        if(childList == null){
           List<Integer> list = new LinkedList<>();
            list.add(key);
            parentList.set(index, list);
        }else{
            if(!childList.contains(key)){
                childList.add(key);
            }
        }
    }
    //we calculate the value of index using the key and get the data from parentlist at that index and then remove the value of that key from childlist
    public void remove(int key) {
        int index = key % array_size;
        List<Integer> childList = parentList.get(index);
        if(childList != null){
            childList.remove(Integer.valueOf(key));
        }
    }
    //it returns true if the childlist is not empty and contains the key
    /** Returns true if this set contains the specified element */
    public boolean contains(int key) {
        int index = key % array_size;
        List<Integer> childList = parentList.get(index);
        return childList != null && childList.contains(key);
    }
}
